// @grpc/grpc-js/build/src/single-subchannel-channel.js
"use strict";
/*
 * Copyright 2025 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleSubchannelChannel = void 0;
const call_number_1 = require("./call-number");
const channelz_1 = require("./channelz");
const compression_filter_1 = require("./compression-filter");
const connectivity_state_1 = require("./connectivity-state");
const constants_1 = require("./constants");
const control_plane_status_1 = require("./control-plane-status");
const deadline_1 = require("./deadline");
const filter_stack_1 = require("./filter-stack");
const metadata_1 = require("./metadata");
const resolver_1 = require("./resolver");
const uri_parser_1 = require("./uri-parser");
class SubchannelCallWrapper {
    constructor(subchannel, method, filterStackFactory, options, callNumber) {
        var _a, _b;
        this.subchannel = subchannel;
        this.method = method;
        this.options = options;
        this.callNumber = callNumber;
        this.childCall = null;
        this.pendingMessage = null;
        this.readPending = false;
        this.halfClosePending = false;
        this.pendingStatus = null;
        this.readFilterPending = false;
        this.writeFilterPending = false;
        const splitPath = this.method.split('/');
        let serviceName = '';
        /* The standard path format is "/{serviceName}/{methodName}", so if we split
          * by '/', the first item should be empty and the second should be the
          * service name */
        if (splitPath.length >= 2) {
            serviceName = splitPath[1];
        }
        const hostname = (_b = (_a = (0, uri_parser_1.splitHostPort)(this.options.host)) === null || _a === void 0 ? void 0 : _a.host) !== null && _b !== void 0 ? _b : 'localhost';
        /* Currently, call credentials are only allowed on HTTPS connections, so we
          * can assume that the scheme is "https" */
        this.serviceUrl = `https://${hostname}/${serviceName}`;
        const timeout = (0, deadline_1.getRelativeTimeout)(options.deadline);
        if (timeout !== Infinity) {
            if (timeout <= 0) {
                this.cancelWithStatus(constants_1.Status.DEADLINE_EXCEEDED, 'Deadline exceeded');
            }
            else {
                setTimeout(() => {
                    this.cancelWithStatus(constants_1.Status.DEADLINE_EXCEEDED, 'Deadline exceeded');
                }, timeout);
            }
        }
        this.filterStack = filterStackFactory.createFilter();
    }
    cancelWithStatus(status, details) {
        if (this.childCall) {
            this.childCall.cancelWithStatus(status, details);
        }
        else {
            this.pendingStatus = {
                code: status,
                details: details,
                metadata: new metadata_1.Metadata()
            };
        }
    }
    getPeer() {
        var _a, _b;
        return (_b = (_a = this.childCall) === null || _a === void 0 ? void 0 : _a.getPeer()) !== null && _b !== void 0 ? _b : this.subchannel.getAddress();
    }
    async start(metadata, listener) {
        if (this.pendingStatus) {
            listener.onReceiveStatus(this.pendingStatus);
            return;
        }
        if (this.subchannel.getConnectivityState() !== connectivity_state_1.ConnectivityState.READY) {
            listener.onReceiveStatus({
                code: constants_1.Status.UNAVAILABLE,
                details: 'Subchannel not ready',
                metadata: new metadata_1.Metadata()
            });
            return;
